### PR TITLE
fix: observe grpc client

### PR DESCRIPTION
**Pull request by gatici**
This PR aims to solve GPRC client connection issues which appears when GPRC server (webconsole) is restarted.

If the Webconsole is restarted, other modules loses the GRPC connection to Webconsole
They tries to connect again for hours with no success.
```
2024-09-26T11:32:10.014Z [smf] 2024-09-26T11:32:10Z [ERRO][Config5g][GRPC] Connectivity status idle, trying to connect again
```

NF observes the GRPC client availability and connection status in an infinite loop.
If the connection is ready it does nothing. If GRPC client does not exist, it creates new client. If GRPC client connection is not ready, it closes the existing client and creates another client.
Only one GRPC client exists for any time slot.

### CODE REVIEW

**Review by gab-arrobo — COMMENTED** [review]
I see that changes are not consistent across the different NFs (in some NFs the `updateConfig` is in `factory/factory.go` like this in NF/PR and in other NFs, the same function goes in `service/init.go`). Would this be a good time to make them consistent?

File: service/init.go [hunk 1]
```diff
@@ -266,8 +266,7 @@ func (smf *SMF) Start() {
 	context.InitSMFUERouting(&factory.UERoutingConfig)
 
 	// Wait for additional/updated config from config pod
-	roc := os.Getenv("MANAGED_BY_CONFIG_POD")
-	if roc == "true" {
+	if os.Getenv("MANAGED_BY_CONFIG_POD") == "true" {
 		initLog.Infof("configuration is managed by Config Pod")
 		initLog.Infof("waiting for initial configuration from config pod")
```
[the remark this answers is not in the document]
Comment (gatici) [reply service/init.go:271]: Done

File: factory/factory.go [hunk 2]
```diff
@@ -46,14 +49,10 @@ func InitConfigFactory(f string) error {
...
 		}
 
-		roc := os.Getenv("MANAGED_BY_CONFIG_POD")
-		if roc == "true" {
-			gClient := client.ConnectToConfigServer(SmfConfig.Configuration.WebuiUri)
-			commChannel := gClient.PublishOnConfigChange(false)
-			go SmfConfig.updateConfig(commChannel)
+		if os.Getenv("MANAGED_BY_CONFIG_POD") == "true" {
+			logger.CfgLog.Infoln("MANAGED_BY_CONFIG_POD is true")
 		}
```
Comment (gab-arrobo) [inline factory/factory.go:54]: Why do you need this `if` statement here given that you are already adding a similar `if` statement in `service/init.go` (same log message "MANAGED_BY_CONFIG_POD is true")?

File: service/init.go [hunk 3]
```diff
@@ -139,9 +141,66 @@ func (smf *SMF) Initialize(c *cli.Context) error {
 		}()
 	}
 
+	if os.Getenv("MANAGED_BY_CONFIG_POD") == "true" {
+		logger.InitLog.Infoln("MANAGED_BY_CONFIG_POD is true")
+		go manageGrpcClient(factory.SmfConfig.Configuration.WebuiUri)
+	}
```
Comment (gab-arrobo) [inline service/init.go:147]: Do not you need to also add/have the `else` statement (similar to the other NFs such as amf, etc.). See this PR: https://github.com/omec-project/amf/pull/330/files#diff-3289814d5eb2b5ebf4f9002fc41f602d2af395ca60676275f6113c0defa8f485

File: factory/factory.go (hunk 2 above)
Comment (gatici) [reply factory/factory.go]: Removed.

File: service/init.go (hunk 3 above)
Comment (gatici) [reply service/init.go]: Done

File: service/init.go [hunk 4]
```diff
@@ -139,9 +141,70 @@ func (smf *SMF) Initialize(c *cli.Context) error {
...
+				stream, err = client.SubscribeToConfigServer()
+				if err != nil {
+					logger.InitLog.Infof("failing SubscribeToConfigServer: %+v", err)
+					continue
+				}
+			}
+
+			if configChannel == nil {
+				configChannel = client.PublishOnConfigChange(true, stream)
+				logger.InitLog.Infoln("PublishOnConfigChange is triggered.")
```
Comment (gab-arrobo) [inline service/init.go:191]:
```suggestion
				logger.InitLog.Infoln("PublishOnConfigChange is triggered")
```

File: service/init.go [hunk 5]
```diff
@@ -139,9 +141,70 @@ func (smf *SMF) Initialize(c *cli.Context) error {
...
+				go factory.SmfConfig.UpdateConfig(configChannel)
+				logger.InitLog.Infoln("SMF updateConfig is triggered.")
```
Comment (gab-arrobo) [inline service/init.go:193]:
```suggestion
				logger.InitLog.Infoln("SMF updateConfig is triggered")
```

File: service/init.go [hunk 6]
```diff
@@ -139,9 +141,70 @@ func (smf *SMF) Initialize(c *cli.Context) error {
...
+			}
+		} else {
+			client, err = grpcClient.ConnectToConfigServer(webuiUri)
+			stream = nil
+			configChannel = nil
+			logger.InitLog.Infoln("connecting to config server.")
```
Comment (gab-arrobo) [inline service/init.go:199]:
```suggestion
			logger.InitLog.Infoln("connecting to config server")
```

File: service/init.go [hunk 7]
```diff
@@ -139,9 +141,70 @@ func (smf *SMF) Initialize(c *cli.Context) error {
...
 	}
 
+	if os.Getenv("MANAGED_BY_CONFIG_POD") == "true" {
+		logger.InitLog.Infoln("MANAGED_BY_CONFIG_POD is true")
+		go manageGrpcClient(factory.SmfConfig.Configuration.WebuiUri)
+	} else {
+		go func() {
+			logger.InitLog.Infoln("use helm chart config")
+			factory.ConfigPodTrigger <- true
+		}()
```
Comment (gab-arrobo) [inline service/init.go:151]: It looks like the `else` statement should not be here (as in the "original" code).
```suggestion
```

When having the `else` statement, most of the tests fail (as you can see below)
```
        "Profile Name: profile4 , Profile Type: uetriggservicereq",
        "Ue's Passed: 1 , Ue's Failed: 4",
        "Profile Errors:",
        "imsi:imsi-208930100007530, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007531, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007532, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007534, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "Profile Status: FAIL",
        "Profile Name: profile2 , Profile Type: pdusessest",
        "Ue's Passed: 0 , Ue's Failed: 5",
        "Profile Errors:",
        "imsi:imsi-208930100007513, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007514, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007510, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007512, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007511, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "Profile Status: FAIL",
        "Profile Name: profile7 , Profile Type: uereqpdusessrelease",
        "Ue's Passed: 0 , Ue's Failed: 5",
        "Profile Errors:",
        "imsi:imsi-208930100007561, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007560, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007562, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007563, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007564, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "Profile Status: FAIL",
        "Profile Name: profile1 , Profile Type: register",
        "Ue's Passed: 0 , Ue's Failed: 5",
        "Profile Errors:",
        "imsi:imsi-208930100007501, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007504, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007500, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007503, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007502, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "Profile Status: FAIL",
        "Profile Name: profile3 , Profile Type: anrelease",
        "Ue's Passed: 0 , Ue's Failed: 5",
        "Profile Errors:",
        "imsi:imsi-208930100007521, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007524, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007520, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007522, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007523, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "Profile Status: FAIL",
        "Profile Name: profile5 , Profile Type: deregister",
        "Ue's Passed: 1 , Ue's Failed: 4",
        "Profile Errors:",
        "imsi:imsi-208930100007544, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007540, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007543, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "imsi:imsi-208930100007541, procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT",
        "Profile Status: FAIL"
```

When I remove the `else` statement (same as in the original code), all tests pass es expected. Did you have a chance to run E2E tests?
```
        "Profile Name: profile1 , Profile Type: register",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS",
        "Profile Name: profile2 , Profile Type: pdusessest",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS",
        "Profile Name: profile3 , Profile Type: anrelease",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS",
        "Profile Name: profile7 , Profile Type: uereqpdusessrelease",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS",
        "Profile Name: profile5 , Profile Type: deregister",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS",
        "Profile Name: profile4 , Profile Type: uetriggservicereq",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS"
```

**Review by gab-arrobo — CHANGES_REQUESTED** [review]

File: service/init.go (hunk 7 above)
Comment (gatici) [reply service/init.go]: I am not able to test else option, but if clause works for me. Anyway, I removed the else clause.

**Review by gab-arrobo — APPROVED** [review]
+1

**Pull request merged**